### PR TITLE
Added the CloudSensei

### PR DESCRIPTION
**Pull request by athiruma**
CloudSensei is an effort to uncover potential cloud resource leaks leading to inefficient cloud management. Currently, we are generating a daily report allowing users to action on long-running EC2 instances.

To-Do:
Eventually, CloudSensei will be flipped to work as a Slackbot, allowing users to join a “read-only” Slack channel to review daily expense reports for instance.

### CODE REVIEW

**Review by krishvoor — CHANGES_REQUESTED** [review]
Overall lgtm, minor nits

File: cloudsensei/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+We have had a few occurrences where cloud resources were 
+running indefinitely even after deleting the cluster. 
+But these few occurrences are good enough to be proactive and 
+this is where “CloudSensei” can give us quick access.
+
+### How it works?
+To implement this functionality, CloudSensei utilizes AWS Lambda, 
+a serverless computing service provided by Amazon Web Services. 
+AWS Lambda allows functions to be triggered without the need for managing 
+a dedicated CI (Continuous Integration) server. By leveraging Lambda,
```
Comment (smalleni) [inline cloudsensei/README.md:15]: I guess the intent here is to mention that with serverless you do not need to manage a dedicated server. Continuous Integration is not relevant in this context. perhaps remove that from this sentence?

File: cloudsensei/lambda_function.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        long_running_instances_by_user = {}
+        for region in regions:
+            region_name = region['RegionName']
+            self.set_ec2_client(region_name)
+            instances_list = self.__ec2_client.describe_instances()['Reservations']
+            for instances in instances_list:
+                for resource in instances['Instances']:
+                    launch_time = resource.get('LaunchTime').date()
+                    days = (current_datetime - launch_time).days
+                    if days > 7:
```
Comment (smalleni) [inline cloudsensei/lambda_function.py:45]: Why don't we make this configurable instead of hardcoding to 7?

File: cloudsensei/run.sh [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    pip install --target ./package -r requirements.txt > $SUCCESS_OUTPUT_PATH
+    pushd package
+    zip -r ../$PROJECT_NAME.zip . > $SUCCESS_OUTPUT_PATH
+    popd
+    zip -g $PROJECT_NAME.zip lambda_function.py > $SUCCESS_OUTPUT_PATH
+
+  pushd terraform
+  echo "#############################"
+  echo "Creating the lambda lambda_function using terraform"
+  if [ -n "$ACCOUNT_ID" ]; then
```
Comment (ebattat) [inline cloudsensei/run.sh:26]: can we use case instead of if-else statement ?

File: cloudsensei/lambda_function.py (hunk 2 above)
Comment (krishvoor) [reply cloudsensei/lambda_function.py:45]: @smalleni we can make it configurable, will default it to 7 days.

File: cloudsensei/run.sh (hunk 3 above)
Comment (athiruma) [reply cloudsensei/run.sh:26]: Here, i was checking whether account_id is exported or not.

File: cloudsensei/agg_lambda/lambda_function.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+from datetime import datetime
+from time import time
+
+from es_operations import ESOperations
+from send_email import send_email_with_ses
+
+
+def lambda_handler(event, context):
+    """
+    This lambda function sends notifications to slack on lon running resources on AWS Cloud
```
Comment (ebattat) [inline cloudsensei/agg_lambda/lambda_function.py:12]: lon ?

File: cloudsensei/agg_lambda/lambda_function.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    :param event:
+    :param context:
+    :return:
+    """
+    start_time = time()
+    print(f"{lambda_handler.__name__} started at {datetime.utcnow()}")
```
Comment (ebattat) [inline cloudsensei/agg_lambda/lambda_function.py:18]: use logger instead of print

File: cloudsensei/env.txt [hunk 6]
```diff
@@ -0,0 +1,13 @@
+ACCOUNT_ID=
```
Comment (ebattat) [inline cloudsensei/env.txt:1]: I think it should be .conf instead of .txt file

File: cloudsensei/lambda_function.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    SLACK_ITEM_SIZE = 50
+
+    def __init__(self):
+        self.__ec2_client = boto3.client('ec2', region_name='us-east-1')
+        self.__iam_client = boto3.client('iam')
+        self.__resource_days = int(os.environ.get('RESOURCE_DAYS', 7))
+
+    def set_ec2_client(self, region_name: str):
+        """
+        This method change the ec2_client object with another region
```
Comment (ebattat) [inline cloudsensei/lambda_function.py:28]: changes

File: cloudsensei/lambda_function.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        :param region_name:
+        :return:
+        """
+        self.__ec2_client = boto3.client('ec2', region_name=region_name)
+
+    def __get_all_instances(self):
+        """
+        This method return all instances in a region
```
Comment (ebattat) [inline cloudsensei/lambda_function.py:36]: returns

File: cloudsensei/es_operations.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import os
+from datetime import datetime
+
+from elasticsearch import Elasticsearch
+
+
+class ESOperations:
```
Comment (ebattat) [inline cloudsensei/es_operations.py:7]: You should add cloud-governance package to requirements.txt and import ESOperations

File: cloudsensei/lambda_function.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+def lambda_handler(event, context):
+    """
+    This lambda function sends notifications to slack on lon running resources on AWS Cloud
+    :param event:
+    :param context:
+    :return:
+    """
+    start_time = time()
+    print(f"{lambda_handler.__name__} started at {datetime.utcnow()}")
```
Comment (ebattat) [inline cloudsensei/lambda_function.py:138]: Use logger instead of print

File: cloudsensei/lambda_function.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        """
+        keys_list = ['User', 'Region', 'Name', 'InstanceId', 'LaunchDate', 'RunningDays']
+        with open('email_template.j2') as template:
+            template = Template(template.read())
+            body = template.render({'resources_list': resources_list, 'keys_list': keys_list})
+            return body
+
+
+def lambda_handler(event, context):
+    """
```
Comment (ebattat) [inline cloudsensei/lambda_function.py:131]: Can you move part of this long method to small methods ?

File: cloudsensei/requirements.txt [hunk 12]
```diff
@@ -0,0 +1,4 @@
+boto3==1.26.1
+requests==2.27.1
+jinja2==3.1.2
+elasticsearch==7.11.0
```
Comment (ebattat) [inline cloudsensei/requirements.txt:4]: add cloud-governance version here

File: cloudsensei/send_email.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
+import ssl
```
Comment (ebattat) [inline cloudsensei/send_email.py:1]: can we use existing module in cloud-governance ?

File: cloudsensei/send_email.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+        msg["To"] = ", ".join(to) if type(to) == list else to
+        if cc:
+            msg["Cc"] = ",".join(cc) if type(cc) == list else cc
+        msg.attach(MIMEText(body, 'html'))
+        try:
+            with SMTP(host, port) as server:
+                server.starttls(context=context)
+                server.login(user=user, password=password)
+                server.send_message(msg)
+                print(f"Successfully sent mail To: {to}, Cc: {cc}")
```
Comment (ebattat) [inline cloudsensei/send_email.py:35]: logger

File: cloudsensei/send_email.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+                return True
+        except Exception as err:
+            print(f"Error raised: {err}")
```
Comment (ebattat) [inline cloudsensei/send_email.py:38]: logger

File: cloudsensei/slack_operations.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
+import os
```
Comment (ebattat) [inline cloudsensei/slack_operations.py:1]: You can add this package to cloud-governance common and import it from there

**Review by ebattat — COMMENTED** [review]
I think that we should shift modules inside cloud-governance package and import it in lambda method

**Comment by athiruma** [comment]
> I think that we should shift modules inside the cloud-governance package and import them in the lambda method

No, this is not a good practice. If we use the cloud-governance package, we need to install all packages installed inside the cloud-governance. 
For using generic packages we need to install the packages locally and upload them to the lambda including the binary packages. 
[Reference](https://docs.aws.amazon.com/lambda/latest/dg/python-package.html#:~:text=zip%20files.-,Creating%20a%20.zip%20deployment%20package%20with%20dependencies,-If%20your%20function)

**Comment by ebattat** [comment]
@athiruma, Yes also there is a size limited

File: cloudsensei/README.md [hunk 17]
```diff
@@ -0,0 +1,91 @@
+## CloudSensei
+
+CloudSensei is a slack bot integrated with cloud governance, 
+allowing users to join a “read-only” slack channel to review 
+daily expense reports for instances (in running state) >= 7 days.
```
Comment (krishvoor) [inline cloudsensei/README.md:5]: Please update to:

CloudSensei is an effort to uncover potential cloud resource leaks leading to inefficient cloud management.  Currently, we are generating a daily report allowing users to action on long-running EC2 instances. 

**To-Do:**
Eventually, CloudSensei will be flipped to work as a Slackbot, allowing users to join a “read-only” Slack channel to review daily expense reports for instance.

**Review by krishvoor — APPROVED** [review]
Minor nit, otherwise lgtm

**Review by ebattat — APPROVED** [review]
/LGTM

**Pull request merged**